core/vm/contracts.go

```diff
@@ -502,3 +526,416 @@ func (c *blake2F) Run(input []byte) ([]byte, error) {
 	}
 	return output, nil
 }
+
+var (
+	errBLS12381InvalidInputLength          = errors.New("invalid input length")
+	errBLS12381InvalidFieldElement         = errors.New("must be less than modulus")
+	errBLS12381InvalidFieldElementTopBytes = errors.New("invalid field element top bytes")
+	errBLS12381G1PointIsNotOnCurve         = errors.New("point is not on curve")
+	errBLS12381G2PointIsNotOnCurve         = errors.New("point is not on curve")
+	errBLS12381G1PointSubgroup             = errors.New("g1 point is not on correct subgroup")
+	errBLS12381G2PointSubgroup             = errors.New("g2 point is not on correct subgroup")
+)
+
+// bls12381G1Add implements EIP-2537 G1Add precompile.
+type bls12381G1Add struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+func (c *bls12381G1Add) RequiredGas(input []byte) uint64 {
+	return params.Bls12381G1AddGas
+}
+
+func (c *bls12381G1Add) Run(input []byte) ([]byte, error) {
+	// Implements EIP-2537 G1Add precompile.
+	// > G1 addition call expects `256` bytes as an input that is interpreted as byte concatenation of two G1 points (`128` bytes each).
+	// > Output is an encoding of addition operation result - single G1 point (`128` bytes).
+	if len(input) != 256 {
+		return nil, errBLS12381InvalidInputLength
+	}
+	var err error
+	var p0, p1 *bls12381.PointG1
+
+	// Initialize G1
+	g := bls12381.NewG1()
```

Comment:
What is the purpose of this `g` variable? Seems to me that all the operations performed "on g" are independent. Wouldn't the code look a lot cleaner are safer if we didn't have this `g` at all?

E.g. `DecodePoint` seems to be fine on its own. Won't tying it to a shared variable make the code more brittle with regard to future changes? It's also extra mental burden, because the reader assumes `g` is used for something, whereas it's not?

Maybe I'm missing something, please provide a bit of clarificarion.

Reply:
`G1`  is the struct that includes all G1 related methods and preallocates for temporary variables. In this particular addition precompile we use G1 not only for decoding but also for _is on curve_ check, performing addition, encoding and converting from Jacobian to affine form.